File: sol/solution/0001-0100/0030. Substring with Concatenation of All Words/Solution.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> res = new ArrayList<>();
        if (words.length == 0 || s.length() == 0) return res;
        int wordLen = words[0].length(), totalLen = wordLen * words.length;
        if (s.length() < totalLen) return res;
        Map<String, Integer> wordCount = new HashMap<>();
        for (String w : words) wordCount.put(w, wordCount.getOrDefault(w, 0) + 1);
        for (int i = 0; i < wordLen; i++) {
            Map<String, Integer> seen = new HashMap<>();
            int left = i, count = 0;
            for (int j = i; j + wordLen <= s.length(); j += wordLen) {
                String w = s.substring(j, j + wordLen);
                if (wordCount.containsKey(w)) {
                    seen.put(w, seen.getOrDefault(w, 0) + 1);
                    count++;
                    while (seen.get(w) > wordCount.get(w)) {
                        String leftWord = s.substring(left, left + wordLen);
                        seen.put(leftWord, seen.get(leftWord) - 1);
                        left += wordLen;
                        count--;
                    }
                    if (count == words.length) res.add(left);
                } else {
                    seen.clear();
                    count = 0;
                    left = j + wordLen;
                }
            }
        }
        return res;
    }
}

File: sol/solution/0001-0100/0030. Substring with Concatenation of All Words/Solution.cs
public class Solution {
    public IList<int> FindSubstring(string s, string[] words) {
        var res = new List<int>();
        if (words.Length == 0 || s.Length == 0) return res;
        int wordLen = words[0].Length, totalLen = wordLen * words.Length;
        if (s.Length < totalLen) return res;
        var wordCount = new Dictionary<string,int>();
        foreach (var w in words) {
            if (!wordCount.ContainsKey(w)) wordCount[w] = 0;
            wordCount[w]++;
        }
        for (int i = 0; i < wordLen; i++) {
            var seen = new Dictionary<string,int>();
            int left = i, count = 0;
            for (int j = i; j + wordLen <= s.Length; j += wordLen) {
                var w = s.Substring(j, wordLen);
                if (wordCount.ContainsKey(w)) {
                    if (!seen.ContainsKey(w)) seen[w] = 0;
                    seen[w]++;
                    count++;
                    while (seen[w] > wordCount[w]) {
                        var leftWord = s.Substring(left, wordLen);
                        seen[leftWord]--;
                        left += wordLen;
                        count--;
                    }
                    if (count == words.Length) res.Add(left);
                } else {
                    seen.Clear();
                    count = 0;
                    left = j + wordLen;
                }
            }
        }
        return res;
    }
}